File: arena-backend/src/modules/user/user.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, IsNull } from 'typeorm';
import { UserEntity } from 'src/entities/user.entity';
import { CreateUserDto } from './dtos/create-user.dto';
import { UpdateUserDto } from './dtos/update-user.dto';
import { WellKnownException } from 'src/exceptions/well-known-exception';
import { generateId } from 'src/utils/id-generator';
import { FileService } from '../file/file.service';
import { Signal } from 'src/signal/signal';
import { SignalChannel } from 'src/signal/signal.channels';

@Injectable()
export class UserService {
  private static readonly UTAG_LENGTH = 6;
  private static readonly UTAG_CHARSET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  private static readonly MAX_UTAG_ATTEMPTS = 10;

  constructor(
    @InjectRepository(UserEntity)
    private readonly userRepository: Repository<UserEntity>,
    private readonly fileService: FileService,
    private readonly signal: Signal,
  ) {}

  async findByUid(uid: string): Promise<UserEntity | null> {
    return this.userRepository.findOne({
      where: { uid, deletedAt: IsNull() },
      relations: ['avatar'],
    });
  }

  async getByUid(uid: string): Promise<UserEntity> {
    const user = await this.findByUid(uid);
    if (!user) {
      throw new WellKnownException({
        message: 'User not found',
        errorCode: 'USER_NOT_FOUND',
      });
    }
    return user;
  }

  async findByUserId(userId: string): Promise<UserEntity | null> {
    return this.userRepository.findOne({
      where: { userId, deletedAt: IsNull() },
      relations: ['avatar'],
    });
  }

  async getByUserId(userId: string): Promise<UserEntity> {
    const user = await this.findByUserId(userId);
    if (!user) {
      throw new WellKnownException({
        message: 'User not found',
        errorCode: 'USER_NOT_FOUND',
      });
    }
    return user;
  }

  async create(uid: string, dto: CreateUserDto): Promise<UserEntity> {
    const existing = await this.findByUid(uid);
    if (existing) {
      throw new WellKnownException({
        message: 'User already exists',
        errorCode: 'ALREADY_EXISTS_USER',
      });
    }

    const utag = await this.generateUniqueUtag();

    const user = this.userRepository.create({
      userId: generateId(),
      uid,
      utag,
      nick: dto.nick,
      email: dto.email ?? null,
      statusMessage: dto.statusMessage ?? null,
      avatarFileId: null,
    });

    await this.userRepository.save(user);

    // 세션 무효화 이벤트 발행
    await this.signal.publish(SignalChannel.USER_UPDATED, { uid });

    // avatar relation 포함해서 조회 (null이지만 일관성 유지)
    return (await this.findByUid(uid))!;
  }

  async update(userId: string, dto: UpdateUserDto): Promise<UserEntity> {
    const user = await this.getByUserId(userId);

    if (dto.nick !== undefined) {
      user.nick = dto.nick;
    }
    if (dto.statusMessage !== undefined) {
      user.statusMessage = dto.statusMessage;
    }
    if (dto.avatarFileId !== undefined) {
      if (dto.avatarFileId !== null) {
        // fileId 존재 여부 검증
        await this.fileService.getFileById(dto.avatarFileId);
      }
      user.avatarFileId = dto.avatarFileId;
    }

    await this.userRepository.save(user);

    // 세션 무효화 이벤트 발행
    await this.signal.publish(SignalChannel.USER_UPDATED, { uid: user.uid });

    // avatar relation 포함해서 다시 조회
    return (await this.findByUserId(userId))!;
  }

  private async generateUniqueUtag(): Promise<string> {
    for (
      let attempt = 0;
      attempt < UserService.MAX_UTAG_ATTEMPTS;
      attempt += 1
    ) {
      const candidate = this.generateRandomUtag();
      const exists = await this.userRepository.exists({
        where: { utag: candidate },
      });
      if (!exists) {
        return candidate;
      }
    }
    throw new WellKnownException({
      message: 'Failed to generate unique utag',
      errorCode: 'UTAG_GENERATION_FAILED',
    });
  }

  private generateRandomUtag(): string {
    let result = '';
    for (let i = 0; i < UserService.UTAG_LENGTH; i += 1) {
      const index = Math.floor(Math.random() * UserService.UTAG_CHARSET.length);
      result += UserService.UTAG_CHARSET.charAt(index);
    }
    return result;
  }
}

File: legacy/backend/src/main/java/app/sandori/arena/api/domain/user/UserService.java
package app.sandori.arena.api.domain.user;

import app.sandori.arena.api.domain.file.FileService;
import app.sandori.arena.api.domain.user.dtos.CreateUserRequest;
import app.sandori.arena.api.domain.user.dtos.UpdateUserRequest;
import app.sandori.arena.api.global.exception.WellKnownException;
import app.sandori.arena.api.global.signal.Signal;
import app.sandori.arena.api.global.signal.SignalChannel;
import app.sandori.arena.api.global.util.IdGenerator;
import java.security.SecureRandom;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserService {

    private static final int UTAG_LENGTH = 6;
    private static final String UTAG_CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    private static final int MAX_UTAG_ATTEMPTS = 10;
    private static final SecureRandom RANDOM = new SecureRandom();

    private final UserRepository userRepository;
    private final FileService fileService;
    private final Signal signal;

    @Transactional(readOnly = true)
    public UserEntity findByUid(String uid) {
        return userRepository.findByUidAndDeletedAtIsNull(uid).orElse(null);
    }

    @Transactional(readOnly = true)
    public UserEntity getByUserId(String userId) {
        return userRepository.findByUserIdAndDeletedAtIsNull(userId)
                .orElseThrow(() -> new WellKnownException("USER_NOT_FOUND", "User not found"));
    }

    @Transactional
    public UserEntity create(String uid, CreateUserRequest dto) {
        UserEntity existing = findByUid(uid);
        if (existing != null) {
            throw new WellKnownException("ALREADY_EXISTS_USER", "User already exists");
        }

        String utag = generateUniqueUtag();

        UserEntity user = UserEntity.builder()
                .userId(IdGenerator.generate())
                .uid(uid)
                .utag(utag)
                .nick(dto.getNick())
                .email(dto.getEmail())
                .statusMessage(dto.getStatusMessage())
                .avatarFileId(null)
                .build();

        userRepository.save(user);

        signal.publish(SignalChannel.USER_UPDATED, Map.of("uid", uid));

        return userRepository.findByUidAndDeletedAtIsNull(uid).orElseThrow();
    }

    @Transactional
    public UserEntity update(String userId, UpdateUserRequest dto) {
        UserEntity user = getByUserId(userId);

        if (dto.getNick() != null) {
            user.setNick(dto.getNick());
        }
        if (dto.getStatusMessage() != null) {
            user.setStatusMessage(dto.getStatusMessage());
        }
        if (dto.isAvatarFileIdProvided()) {
            if (dto.getAvatarFileId() != null) {
                fileService.getFileById(dto.getAvatarFileId());
            }
            user.setAvatarFileId(dto.getAvatarFileId());
        }

        userRepository.save(user);

        signal.publish(SignalChannel.USER_UPDATED, Map.of("uid", user.getUid()));

        return userRepository.findByUserIdAndDeletedAtIsNull(userId).orElseThrow();
    }

    private String generateUniqueUtag() {
        for (int attempt = 0; attempt < MAX_UTAG_ATTEMPTS; attempt++) {
            String candidate = generateRandomUtag();
            if (!userRepository.existsByUtag(candidate)) {
                return candidate;
            }
        }
        throw new WellKnownException("UTAG_GENERATION_FAILED", "Failed to generate unique utag");
    }

    private String generateRandomUtag() {
        StringBuilder sb = new StringBuilder(UTAG_LENGTH);
        for (int i = 0; i < UTAG_LENGTH; i++) {
            sb.append(UTAG_CHARSET.charAt(RANDOM.nextInt(UTAG_CHARSET.length())));
        }
        return sb.toString();
    }
}
